src/main/kotlin/com/zepben/evolve/cim/iec61970/base/core/NameType.kt

```diff
@@ -55,23 +76,28 @@ class NameType(val name: String) {
                     existing
                 } else {
                     Name(name, this, identifiedObject).also {
-                        namesMultiIndex[name] = mutableListOf(existing, it)
+                        namesMultiIndex[name] = mutableMapOf(existing.identifiedObject to existing, it.identifiedObject to it)
                         namesIndex.remove(name)
+                        identifiedObject.addName(this, name)
                     }
                 }
             }
+
             namesMultiIndex.containsKey(name) -> {
                 val names = namesMultiIndex[name]!!
-                var nameObj = names.find { it.identifiedObject == identifiedObject }
+                var nameObj = names[identifiedObject]
                 if (nameObj == null) {
                     nameObj = Name(name, this, identifiedObject)
-                    names.add(nameObj)
+                    names[identifiedObject] = nameObj
+                    identifiedObject.addName(this, name)
                 }
                 nameObj
```

Comment:
todo: This is very Java-esk. Can you please update it to a more idiomatic Kotlin. e.g.
```
names[identifiedObject] ?: Name(name, this, identifiedObject).also {
    names[identifiedObject] = it
    identifiedObject.addName(this, name)
}
```

Reply:
Done